File: TinyScript/java/src/main/java/parser/SimpleParser.java
package parser;

import parser.ast.*;
import parser.util.ParseException;
import parser.util.PeekTokenIterator;

public class SimpleParser {
    // Expr -> digit + Expr | digit
    // digit -> 0|1|2|3|4|5|...|9
    public static ASTNode parse(PeekTokenIterator it) throws ParseException {

        var expr = new Expr();
        var scalar = Factor.parse(it);
        // base condition
        if(!it.hasNext()){
           return scalar;
        }

        expr.setLexeme(it.peek());
        it.nextMatch("+");
        expr.setLabel("+");
        expr.addChild(scalar);
        expr.setType(ASTNodeTypes.BINARY_EXPR);
        var rightNode = parse(it);
        expr.addChild(rightNode);
        return expr;
    }
}

File: TinyScript/js/src/parser/SimpleParser.js
const Expr = require("./ast/Expr")
const Factor = require('./ast/Factor')
const ASTNodeTypes = require('./ast/ASTNodeTypes')

class SimpleParser {

    // Expr -> digit + Expr | digit
    // digit -> 0|1|2|3|...|9
    static parse(it) {
        const expr = new Expr(null)
        const scalar = Factor.parse(it)
        if(!it.hasNext()) {
            return scalar
        }

        expr.addChild(scalar)
        const op = it.nextMatch("+")
        expr.label = "+"
        expr.type = ASTNodeTypes.BINARY_EXPR
        expr.lexeme = op
        expr.addChild(SimpleParser.parse(it))

        return expr
    }
}

module.exports = SimpleParser